foreman_stub.py

```diff
@@ -9,61 +9,146 @@
-        pagenum = int(pagenum)
-    return pagenum
+    """Return page number."""
+    return int(request.args.get('page', 1))
+
+
+def get_per_page():
+    """Return the number of results per page."""
+    return int(request.args.get('per_page', 20))
+
+
+def find_host(hostid):
+    try:
+        return PAGECACHE[f'/hosts/{hostid}'][1]
+    except KeyError:
+        abort(404)
 
 
 @app.route('/api/v2/hosts')
-@app.route('/api/v2/hosts/<hostid>')
-@app.route('/api/v2/hosts/<hostid>/facts')
-def get_hosts(hostid=None):
+def get_hosts():
     """Render fixture contents from cache."""
     pagenum = get_page_num()
+    resp = PAGECACHE['/hosts'][pagenum]
+    return jsonify(resp)
 
-    cache_key = '/hosts'
-    if hostid is not None:
-        cache_key = f'{cache_key}/{hostid}'
 
-    resp = PAGECACHE[cache_key]
+@app.route('/api/v2/hosts/facts', methods=['POST'])
+def handle_facts():
+    data = request.get_json()
+
+    # Foreman can allow creating hosts that don't exist - we don't support this
+    host = find_host(data['name'])
+
+    host['facts'] = data['facts']
+
+    return jsonify(host), 201
+
+
+@app.route('/api/v2/hosts/<hostid>')
+def get_host(hostid):
+    """Render fixture contents from cache."""
+    return jsonify(find_host(hostid))
+
+
+@app.route('/api/v2/hosts/<hostid>/facts', methods=['GET'])
+def get_facts(hostid):
+    host = find_host(hostid)
+
+    if request.method == 'POST':
+        host['facts'] = request.get_json()
+
+    page = get_page_num()
+    per_page = get_per_page()
+
+    start = (page - 1) * per_page
+    end = (page) * per_page
+
+    resp = {
+        'total': len(host['facts']),
+        'subotal': 1,  # this appears to always be 1 in the API
+        'page': page,
+        'per_page': per_page,
+        'sort': {
+            'by': None,
+            'order': None,
+        },
+        'results': {
+            host['name']: dict(sorted(host['facts'].items())[start:end]),
+        },
+    }
+
+    return jsonify(resp)
+
+
+@app.route('/api/v2/config_reports', methods=['GET'])
+def get_config_reports():
+    page = get_page_num()
+    per_page = get_per_page()
+
+    start = (page - 1) * per_page
+    end = (page) * per_page
+
+    resp = {
+        'total': len(REPORTS),
+        'subotal': len(REPORTS),
+        'page': page,
+        'per_page': per_page,
+        'sort': {
+            'by': None,
+            'order': None,
+        },
+        'results': REPORTS[start:end],
+    }
+
+    return jsonify(resp)
+
+@app.route('/api/v2/config_reports', methods=['POST'])
+def create_config_report():
+    data = request.get_json()
     try:
-        if not request.path.endswith('/facts'):
-            resp = resp[pagenum]
+        report = data['config_report']
     except KeyError:
-        if DEBUG:
-            import q; q/cache_key; q/pagenum; q/hostid
-            import epdb; epdb.st()
-    else:
-        return jsonify(resp)
+        abort(400)
+
+    REPORTS.append(report)
+    resp = {}
```

Comment:
Extra var is definitely not needed here

Reply:
This is why it's a draft: I still need to look into the exact response.